### PR TITLE
create basic contribution guidelines

### PR DESCRIPTION
**Pull request by LemmaEOF**
This is very non-final, but I feel like we should have a base set of guidelines for how to name things and contribute to Yarn. Feedback is appreciated.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+contributors, since people already have a conception of what a method "should" be named. Because of this, we have decided to 
+take a clean-room approach with mapping Minecraft to avoid potential legal/drama issues within the community.
+
+## What's in a Name?
+
+The fundamental goal of mappings is to make modding Minecraft easier to understand, for beginners and experts alike. We want
+to create a cohesive, simple naming scheme that avoids confusion or being too verbose. Having a lot of things named in
+different schemes would be a nightmare for usability, so we want to avoid that. This includes naming schemes not just within
+Yarn but within the greater Minecraft community. Despite that goal, there are some name sources that are inherently more
+"reliable" than others; you should use an up-to-date and known Mojang name above any alternative, even if an alternative
```
Comment (3TUSK) [inline CONTRIBUTING.md]: If I recall right, Mojang calls `Identifier` as `NamespacedString` - which means that there would be exceptions to this sentence.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+terrible choices for names) sticking around. While it would be nice to have names like `Gui` or the like, we have decided to 
+use `Screen`, since that's the name for guis that the client-side registry uses. Mojang names can be found in many places,
+from registry names to lang files to tweets by devs. If you want to claim something as a Mojang name, and it's not an obvious
+one, provide a source for proof.
+
+Minecraft is still under active development, and names have been chaned in the past. The most notable case of this is `Tile`,
+which used to be the name for Blocks. This has changed in 1.13 and 1.14, so we use the modern name `BlockEntity` instead of
+the old Notch name `TileEntity`.
+
+### 2. Avoid MCP names wherever possible.
```
Comment (3TUSK) [inline CONTRIBUTING.md]: Though not being a real situation, I would word this title defensively - that said, I would say "Avoid MCP names and other proprietary names wherever possible".

**Review by 3TUSK — COMMENTED** [review]
Some comments. Also, should we mention some code-style related rules (CAPITALIZE_CONSTANTS_NAMES, camelCase for method names, etc.)?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Contributing to Yarn
+
+Yarm is in a slightly precarious position, due to the odd nature of MCP. MCP is all-rights-reserved, yet visible to anyone
```
Comment (falkreon) [inline CONTRIBUTING.md]: Yarm->yarn

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+### 1. Use modern Mojang names when known.
+
+This can be a contentious decision, due to many of Notch's original names for things (that he himself has stated were
+terrible choices for names) sticking around. While it would be nice to have names like `Gui` or the like, we have decided to 
+use `Screen`, since that's the name for guis that the client-side registry uses. Mojang names can be found in many places,
+from registry names to lang files to tweets by devs. If you want to claim something as a Mojang name, and it's not an obvious
+one, provide a source for proof.
+
+Minecraft is still under active development, and names have been chaned in the past. The most notable case of this is `Tile`,
```
Comment (falkreon) [inline CONTRIBUTING.md]: chaned->changed

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+settled on a name scheme that involves abbreviations. Abbreviations should be avoided when possible, but when you have
+classes with names like `ScoreboardObjectiveUpdateS2CPacket`, the abbreviations are necessary to prevent the name being
+even more excessively long.
+
+### 5. Don't be afraid to dispute existing names.
+
+The Fabric community is active and evolving. Because of this, what's agreed upon as a good name can change over time. If you
+have an issue with a name and don't feel like it fits in Yarn, don't hesitate to make an issue about it. We're still having
+active discussion about packet naming schemes at the time of writing, and we may choose to change the scheme in the future if
+enoug people think a change is warranted. We want to avoid a LexManos situation where one person has to rule with an iron
```
Comment (falkreon) [inline CONTRIBUTING.md]: enoug->enough

File: CONTRIBUTING.md (hunk 1 above)
Comment (falkreon) [reply CONTRIBUTING.md]: So maybe "clear intent" > "official/modern mojang" when the difference is great, and the decision on this is made by consensus (or, in extreme cases, fearless leader decree)?

**Review by falkreon — CHANGES_REQUESTED** [review]
Minor spelling things

**Review by falkreon — APPROVED** [review]
All set!

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
+# Contributing to Yarn
+
+Yarn is in a slightly precarious position, due to the odd nature of MCP. MCP is all-rights-reserved, yet visible to anyone
```
Comment (asiekierka) [inline CONTRIBUTING.md]: Drop the MCP stuff from CONTRIBUTING.md, IMO. I think the big thick warning in README.md is enough.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+contributors, since people already have a conception of what a method "should" be named. Because of this, we have decided to
+take a clean-room approach with mapping Minecraft to avoid potential legal/drama issues within the community.
+
+## What's in a Name?
+
+The fundamental goal of mappings is to make modding Minecraft easier to understand, for beginners and experts alike. We want
+to create a cohesive, simple naming scheme that avoids confusion or being too verbose. Having a lot of things named in
+different schemes would be a nightmare for usability, so we want to avoid that. This includes naming schemes not just within
+Yarn but within the greater Minecraft community. Despite that goal, there are some name sources that are inherently more
+"reliable" than others; you should use an up-to-date and known Mojang name above any alternative, even if an alternative
```
Comment (asiekierka) [inline CONTRIBUTING.md]: "you should use an up-to-date and known Mojang name [...] be simpler." -> "if an official Mojang name is hinted at or can be guessed, you should generally use that - sticking closer to official terminology helps everyone."

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+would be simpler.
+
+## Guidelines
+
+To keep with the goal of consistency within the mappings and the greater Minecraft community, here are a set of guidelines to
+follow when naming unmapped methods.
+
+### 1. Use modern, sensible Mojang names when known.
```
Comment (asiekierka) [inline CONTRIBUTING.md]: 1. Stick to official terminology.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+This can be a contentious decision, due to many of Notch's original names for things (that he himself has stated were
+terrible choices for names) sticking around. While it would be nice to have names like `Gui` or the like, we have decided to
+use `Screen`, since that's the name for guis that the client-side registry uses. Mojang names can be found in many places,
```
Comment (asiekierka) [inline CONTRIBUTING.md]: Mojang names -> official terms

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+from registry names to lang files to tweets by devs. If you want to claim something as a Mojang name, and it's not an obvious
+one, provide a source for proof.
```
Comment (asiekierka) [inline CONTRIBUTING.md]: + In addition, some things - like enum names or registered object fields - have automatically-derived names; generally, those should not be changed.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+Minecraft is still under active development, and names have been changed in the past. The most notable case of this is `Tile`,
+which used to be the name for Blocks. This has changed in 1.13 and 1.14, so we use the modern name `BlockEntity` instead of
+the old Notch name `TileEntity`.
```
Comment (asiekierka) [inline CONTRIBUTING.md]: the older name

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+There are a few situations when we've decided that Mojang's names are too obtuse or clunky for our mappings, where we've
+decided to use our own names. The most obvious example is our use of `Identifier` when Mojang uses `NamespacedString`. In
```
Comment (asiekierka) [inline CONTRIBUTING.md]: uses -> is assumed to use

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+terrible choices for names) sticking around. While it would be nice to have names like `Gui` or the like, we have decided to
+use `Screen`, since that's the name for guis that the client-side registry uses. Mojang names can be found in many places,
+from registry names to lang files to tweets by devs. If you want to claim something as a Mojang name, and it's not an obvious
+one, provide a source for proof.
+
+Minecraft is still under active development, and names have been changed in the past. The most notable case of this is `Tile`,
+which used to be the name for Blocks. This has changed in 1.13 and 1.14, so we use the modern name `BlockEntity` instead of
+the old Notch name `TileEntity`.
+
+There are a few situations when we've decided that Mojang's names are too obtuse or clunky for our mappings, where we've
```
Comment (asiekierka) [inline CONTRIBUTING.md]: Mojang's names -> official names

**Comment by asiekierka** [comment]
Not set at all. I'll go patch this up - some wording is tricky...

**Comment by asiekierka** [comment]
https://github.com/FabricMC/yarn/pull/477

Done a take two, as I just stopped commenting halfway through... Could we move discussion to that?

**Comment by LemmaEOF** [comment]
I'm happy with the new version from Asie, so I'm gonna close this down.

**Pull request closed without merging**